drivers/meter/ade9430/ade9430.c

```diff
@@ -0,0 +1,325 @@
+{
+	uint8_t buff[6] = {0};
+
+	buff[0] = reg_addr >> 4;
+	buff[1] = reg_addr << 4;
+
+	if (reg_addr >= ADE9430_REG_RUN && reg_addr <= ADE9430_REG_VERSION) {
+		no_os_put_unaligned_be16(reg_data, &buff[2]);
+		return no_os_spi_write_and_read(dev->spi_desc, buff, 4);
+	}
+
+	no_os_put_unaligned_be32(reg_data, &buff[2]);
+	return no_os_spi_write_and_read(dev->spi_desc, buff, 6);
+}
+
+/**
+ * @brief Update specific register bits.
+ * @param dev - The device structure.
+ * @param reg_addr - The register address.
+ * @param mask - Specific bits mask.
+ * @param reg_data - The data to be written.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int ade9430_update_bits(struct ade9430_dev *dev, uint16_t reg_addr,
+			uint32_t mask, uint32_t reg_data)
+{
+	int ret;
+	uint32_t data;
+
+	ret = ade9430_read(dev, reg_addr, &data);
+	if (ret)
+		return ret;
+
+	data &= ~mask;
+	data |= reg_data & mask;
+
+	return ade9430_write(dev, reg_addr, data);
+}
+
+/**
+ * @brief Initialize the device.
+ * @param device - The device structure.
+ * @param init_param - The structure that contains the device initial
+ * 		       parameters.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int ade9430_init(struct ade9430_dev **device,
+		 struct ade9430_init_param init_param)
+{
+	struct ade9430_dev *dev;
+	uint32_t chip_id;
+	int ret;
+
+	dev = (struct ade9430_dev *)calloc(1, sizeof(*dev));
+	if (!dev)
+		return -ENOMEM;
+
+	/* SPI Initialization*/
+	ret = no_os_spi_init(&dev->spi_desc, init_param.spi_init);
+	if (ret)
+		goto error_dev;
+
+	ret = ade9430_update_bits(dev, ADE9430_REG_CONFIG1, ADE9430_SWRST,
+				  no_os_field_prep(ADE9430_SWRST, 1));
+	if (ret)
+		goto error_spi;
+
+	no_os_mdelay(100);
+
+	ret = ade9430_update_bits(dev, ADE9430_REG_CONFIG1, ADE9430_SWRST,
+				  no_os_field_prep(ADE9430_SWRST, 0));
+	if (ret)
+		goto error_spi;
+
+	/** Use a valid register with default value different from 0*/
+	ret = ade9430_read(dev, ADE9430_REG_CONFIG5, &chip_id);
+	if (ret)
+		goto error_spi;
+
+	ret = ade9430_write(dev, ADE9430_REG_RUN, 0);
+	if (ret)
+		goto error_spi;
+
+	if (chip_id != ADE9430_CHIP_ID)
+		goto error_spi;
+
+	/** Accumulate Energy over 7999 samples (1 second accumulation) */
+	ret = ade9430_write(dev, ADE9430_REG_EP_CFG,
+			    no_os_field_prep(ADE9430_EGY_LD_ACCUM, 1) |
+			    no_os_field_prep(ADE9430_EGY_TMR_MODE, 0) |
+			    no_os_field_prep(ADE9430_RD_RST_EN, 0) |
+			    no_os_field_prep(ADE9430_EGY_PWR_EN, 1));
+	if (ret)
+		goto error_spi;
+
+	/* 1 sec accumulation */
+	ret = ade9430_write(dev, ADE9430_REG_EGY_TIME, 7999);
+	if (ret)
+		goto error_spi;
+
+	ret = ade9430_write(dev, ADE9430_REG_PWR_TIME, 7999);
+	if (ret)
+		goto error_spi;
+
+	/* Enable Temperature */
+	ret = ade9430_update_bits(dev, ADE9430_REG_TEMP_CFG, ADE9430_TEMP_EN,
+				  no_os_field_prep(ADE9430_TEMP_EN, 1));
+	if (ret)
+		goto error_spi;
+
+	ret = ade9430_write(dev, ADE9430_REG_RUN, 1);
+	if (ret)
+		goto error_spi;
+
```

Comment:
So this whole initialization sequence above seems tailored for a specific project because the drive hardcodes those values.

Device driver should be "project-agnostic" and allow for configuration of these things through init params to be able to fit any project.

Reply:
Indeed, in this driver there is only one implementation of the three User Energy Register Use Models that are available at:
https://wiki.analog.com/resources/eval/user-guides/ade9430

I can add support for the other 2 also.